File: backend/compartido/dominio/validadores.py
"""
validadores.py — Validación pura del grafo (port desde frontend tipos.ts)
"""

from .modelos import Nodo, Arista, Red, ErrorValidacion


TIPOS_NODO = {"AND", "OR", "NOT"}


def validar_nodo(nodo: Nodo, ids_existentes: set[str] | None = None) -> list[ErrorValidacion]:
    errores: list[ErrorValidacion] = []

    if not nodo.id:
        errores.append(ErrorValidacion(codigo="NODO_SIN_ID", mensaje="El nodo no tiene identificador."))
    elif not nodo.id.startswith("nodo-"):
        errores.append(ErrorValidacion(
            codigo="NODO_FORMATO_INVALIDO",
            mensaje=f'El identificador "{nodo.id}" debe seguir el formato "nodo-*".',
            nodo_id=nodo.id,
        ))

    if ids_existentes and nodo.id in ids_existentes:
        errores.append(ErrorValidacion(
            codigo="NODO_DUPLICADO",
            mensaje=f'Ya existe un nodo con id "{nodo.id}".',
            nodo_id=nodo.id,
        ))

    if not nodo.etiqueta or not nodo.etiqueta.strip():
        errores.append(ErrorValidacion(
            codigo="NODO_SIN_NOMBRE",
            mensaje=f'El nodo "{nodo.id}" no tiene etiqueta.',
            nodo_id=nodo.id,
        ))

    if nodo.tipo not in TIPOS_NODO:
        errores.append(ErrorValidacion(
            codigo="TIPO_NODO_INVALIDO",
            mensaje=f'El tipo "{nodo.tipo}" debe ser uno de: {", ".join(TIPOS_NODO)}.',
            nodo_id=nodo.id,
        ))

    if not isinstance(nodo.propiedades, dict):
        errores.append(ErrorValidacion(
            codigo="PROPIEDADES_INVALIDAS",
            mensaje=f'El nodo "{nodo.id}" tiene propiedades que no son un objeto.',
            nodo_id=nodo.id,
        ))

    if not _es_finito(nodo.posicion.x) or not _es_finito(nodo.posicion.y):
        errores.append(ErrorValidacion(
            codigo="POSICION_INVALIDA",
            mensaje=f'El nodo "{nodo.id}" tiene coordenadas inválidas.',
            nodo_id=nodo.id,
        ))

    return errores


def validar_arista(
    arista: Arista,
    ids_nodos: set[str],
    ids_aristas: set[str] | None = None,
) -> list[ErrorValidacion]:
    errores: list[ErrorValidacion] = []

    if not arista.id:
        errores.append(ErrorValidacion(codigo="ARISTA_SIN_ID", mensaje="La arista no tiene identificador."))
    elif not arista.id.startswith("arista-"):
        errores.append(ErrorValidacion(
            codigo="ARISTA_FORMATO_INVALIDO",
            mensaje=f'El identificador "{arista.id}" debe seguir el formato "arista-*".',
            arista_id=arista.id,
        ))

    if ids_aristas and arista.id in ids_aristas:
        errores.append(ErrorValidacion(
            codigo="ARISTA_DUPLICADA",
            mensaje=f'Ya existe una arista con id "{arista.id}".',
            arista_id=arista.id,
        ))

    if not ids_nodos or arista.id_origen not in ids_nodos:
        errores.append(ErrorValidacion(
            codigo="ORIGEN_INEXISTENTE",
            mensaje=f'La arista "{arista.id}" referencia un nodo origen inexistente.',
            arista_id=arista.id,
        ))

    if not ids_nodos or arista.id_destino not in ids_nodos:
        errores.append(ErrorValidacion(
            codigo="DESTINO_INEXISTENTE",
            mensaje=f'La arista "{arista.id}" referencia un nodo destino inexistente.',
            arista_id=arista.id,
        ))

    if arista.id_origen == arista.id_destino:
        errores.append(ErrorValidacion(
            codigo="AUTO_LOOP",
            mensaje=f'La arista "{arista.id}" no puede conectar un nodo consigo mismo.',
            arista_id=arista.id,
        ))

    if not _es_finito(arista.peso):
        errores.append(ErrorValidacion(
            codigo="PESO_INVALIDO",
            mensaje=f'La arista "{arista.id}" tiene un peso inválido: {arista.peso}.',
            arista_id=arista.id,
        ))

    return errores


def validar_matriz(matriz) -> list[ErrorValidacion]:
    errores: list[ErrorValidacion] = []

    if not hasattr(matriz, "id") or not matriz.id:
        errores.append(ErrorValidacion(codigo="MATRIZ_SIN_ID", mensaje="La matriz no tiene identificador."))

    if not hasattr(matriz, "valores") or not matriz.valores:
        errores.append(ErrorValidacion(codigo="MATRIZ_VACIA", mensaje=f"La matriz no tiene valores."))
        return errores

    filas = matriz.valores
    cols = len(filas[0]) if filas else 0

    if cols == 0:
        errores.append(ErrorValidacion(
            codigo="MATRIZ_VACIA",
            mensaje=f"La matriz tiene 0 columnas.",
        ))
        return errores

    for i, fila in enumerate(filas):
        if len(fila) != cols:
            errores.append(ErrorValidacion(
                codigo="MATRIZ_FILA_HETEROGENEA",
                mensaje=f"La fila {i} tiene {len(fila)} columnas, se esperaban {cols}.",
            ))

    for i, fila in enumerate(filas):
        for j, valor in enumerate(fila):
            if not isinstance(valor, (int, float)) or not _es_finito(valor):
                errores.append(ErrorValidacion(
                    codigo="MATRIZ_VALOR_INVALIDO",
                    mensaje=f"El valor en [{i}][{j}] no es numérico o es infinito.",
                ))

    return errores


def validar_red(red: Red) -> list[ErrorValidacion]:
    errores: list[ErrorValidacion] = []

    if not red.id:
        errores.append(ErrorValidacion(codigo="RED_SIN_ID", mensaje="La red no tiene identificador."))

    if not red.nodos or len(red.nodos) == 0:
        errores.append(ErrorValidacion(codigo="GRAFO_VACIO", mensaje="La red no contiene nodos."))
        return errores

    ids_nodos: set[str] = set()
    ids_aristas: set[str] = set()

    for nodo in red.nodos:
        errores_nodo = validar_nodo(nodo, ids_nodos)
        errores.extend(errores_nodo)
        if nodo.id and nodo.id not in ids_nodos:
            ids_nodos.add(nodo.id)

    for arista in red.aristas:
        errores_arista = validar_arista(arista, ids_nodos, ids_aristas)
        errores.extend(errores_arista)
        if arista.id and arista.id not in ids_aristas:
            ids_aristas.add(arista.id)

    if red.aristas and _detectar_ciclos(red.nodos, red.aristas):
        errores.append(ErrorValidacion(
            codigo="CICLO_DETECTADO",
            mensaje="La red contiene uno o más ciclos dirigidos.",
        ))

    return errores


def _detectar_ciclos(nodos: list[Nodo], aristas: list[Arista]) -> bool:
    ady: dict[str, list[str]] = {n.id: [] for n in nodos}
    for a in aristas:
        if a.id_origen in ady:
            ady[a.id_origen].append(a.id_destino)

    color: dict[str, int] = {n.id: 0 for n in nodos}  # 0=blanco,1=gris,2=negro

    def dfs(nid: str) -> bool:
        color[nid] = 1
        for v in ady.get(nid, []):
            c = color.get(v, 0)
            if c == 1:
                return True
            if c == 0 and dfs(v):
                return True
        color[nid] = 2
        return False

    for n in nodos:
        if color.get(n.id, 0) == 0 and dfs(n.id):
            return True

    return False


def _es_finito(valor) -> bool:
    return isinstance(valor, (int, float)) and float(valor) != float('inf') and valor == valor

File: frontend/src/compartido/dominio/validadores.ts
import type {
  Nodo, Arista, Red, Conectivo, Matriz,
  ErrorValidacion, Id, ValorEvidencia, OperacionLogica,
} from './tipos';

/**
 * ============================================================
 * Prompt: "Crea funciones de validación puras para Nodo,
 * Arista, Red (con detección de ciclos DFS), Conectivo y
 * Matriz. Reutilizable frontend/backend."
 * ============================================================
 * validadores.ts — Validación Pura del Grafo
 * ============================================================
 * Funciones sin efectos secundarios. Reutilizables frontend/backend.
 */

// ─── Helpers ────────────────────────────────────────────────

type ColorVisita = 'blanco' | 'gris' | 'negro';

const VALORES_EVIDENCIA: readonly ValorEvidencia[] = ['T', 'F', 'B', 'N'];
const OPERACIONES_LOGICAS: readonly OperacionLogica[] = ['AND', 'OR', 'XOR', 'NOR', 'NAND', 'NOT'];

function detectarCiclos(nodos: readonly Nodo[], aristas: readonly Arista[]): boolean {
  const ady = new Map<Id, Id[]>();
  for (const n of nodos) ady.set(n.id, []);
  for (const a of aristas) ady.get(a.origenId)?.push(a.destinoId);

  const color = new Map<Id, ColorVisita>();
  for (const n of nodos) color.set(n.id, 'blanco');

  function dfs(id: Id): boolean {
    color.set(id, 'gris');
    for (const v of ady.get(id) ?? []) {
      const c = color.get(v);
      if (c === 'gris') return true;
      if (c === 'blanco' && dfs(v)) return true;
    }
    color.set(id, 'negro');
    return false;
  }

  for (const n of nodos) {
    if (color.get(n.id) === 'blanco' && dfs(n.id)) return true;
  }
  return false;
}

// ─── Valor evidencial ───────────────────────────────────────

export function validarValorEvidencia(valor: string, campo: string, id?: Id): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];
  if (!VALORES_EVIDENCIA.includes(valor as ValorEvidencia)) {
    errores.push({
      codigo: 'VALOR_EVIDENCIA_INVALIDO',
      mensaje: `"${campo}" debe ser uno de: ${VALORES_EVIDENCIA.join(', ')}. Recibido: "${valor}".`,
      nodoId: id,
    });
  }
  return errores;
}

// ─── validarNodo ────────────────────────────────────────────

export function validarNodo(
  nodo: Nodo,
  idsExistentes?: Set<Id>,
): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];

  if (!nodo.id) errores.push({ codigo: 'NODO_SIN_ID', mensaje: 'El nodo no tiene identificador.' });
  if (!nodo.etiqueta?.trim()) errores.push({ codigo: 'NODO_SIN_NOMBRE', mensaje: `El nodo "${nodo.id}" no tiene etiqueta.`, nodoId: nodo.id });
  if (typeof nodo.peso !== 'number' || !Number.isFinite(nodo.peso)) errores.push({ codigo: 'PESO_INVALIDO', mensaje: `Nodo "${nodo.id}" peso inválido: ${nodo.peso}.`, nodoId: nodo.id });
  if (!Number.isFinite(nodo.posicion?.x) || !Number.isFinite(nodo.posicion?.y)) errores.push({ codigo: 'POSICION_INVALIDA', mensaje: `Nodo "${nodo.id}" coordenadas inválidas.`, nodoId: nodo.id });
  if (idsExistentes?.has(nodo.id)) errores.push({ codigo: 'NODO_DUPLICADO', mensaje: `Ya existe nodo con id "${nodo.id}".`, nodoId: nodo.id });

  if (nodo.tipo === 'operacion') {
    if (!nodo.operacion) {
      errores.push({ codigo: 'OPERACION_FALTANTE', mensaje: `Nodo "${nodo.id}" es de tipo operación pero no tiene operación asignada.`, nodoId: nodo.id });
    } else if (!OPERACIONES_LOGICAS.includes(nodo.operacion)) {
      errores.push({ codigo: 'OPERACION_INVALIDA', mensaje: `Nodo "${nodo.id}" operación inválida: "${nodo.operacion}".`, nodoId: nodo.id });
    }
  } else if (nodo.tipo === 'evidencia') {
    errores.push(...validarValorEvidencia(nodo.valorInicial, 'valorInicial', nodo.id));
    errores.push(...validarValorEvidencia(nodo.valorActual, 'valorActual', nodo.id));
  }

  if (typeof nodo.tamano !== 'number' || nodo.tamano <= 0) {
    errores.push({ codigo: 'TAMANO_INVALIDO', mensaje: `Nodo "${nodo.id}" tamaño inválido: ${nodo.tamano}.`, nodoId: nodo.id });
  }

  if (typeof nodo.color !== 'string' || !nodo.color.startsWith('#')) {
    errores.push({ codigo: 'COLOR_INVALIDO', mensaje: `Nodo "${nodo.id}" color inválido: "${nodo.color}". Debe ser hex.`, nodoId: nodo.id });
  }

  return errores;
}

// ─── validarArista ──────────────────────────────────────────

export function validarArista(
  arista: Arista,
  idsNodos: Set<Id>,
  idsAristas?: Set<Id>,
  nodos?: readonly Nodo[],
): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];

  if (!arista.id) errores.push({ codigo: 'ARISTA_SIN_ID', mensaje: 'La arista no tiene identificador.' });
  if (!idsNodos.has(arista.origenId)) errores.push({ codigo: 'REFERENCIA_ORIGEN_INEXISTENTE', mensaje: `Arista "${arista.id}" referencia origen "${arista.origenId}" inexistente.`, aristaId: arista.id });
  if (!idsNodos.has(arista.destinoId)) errores.push({ codigo: 'REFERENCIA_DESTINO_INEXISTENTE', mensaje: `Arista "${arista.id}" referencia destino "${arista.destinoId}" inexistente.`, aristaId: arista.id });
  if (typeof arista.peso !== 'number' || !Number.isFinite(arista.peso)) errores.push({ codigo: 'PESO_ARISTA_INVALIDO', mensaje: `Arista "${arista.id}" peso inválido: ${arista.peso}.`, aristaId: arista.id });
  if (idsAristas?.has(arista.id)) errores.push({ codigo: 'ARISTA_DUPLICADA', mensaje: `Ya existe arista con id "${arista.id}".`, aristaId: arista.id });
  if (typeof arista.grosor !== 'number' || arista.grosor <= 0) errores.push({ codigo: 'GROSOR_INVALIDO', mensaje: `Arista "${arista.id}" grosor inválido.`, aristaId: arista.id });
  if (!arista.color?.startsWith('#')) errores.push({ codigo: 'COLOR_ARISTA_INVALIDO', mensaje: `Arista "${arista.id}" color inválido.`, aristaId: arista.id });

  if (nodos) {
    const origen = nodos.find((n) => n.id === arista.origenId);
    const destino = nodos.find((n) => n.id === arista.destinoId);
    if (origen && destino && origen.tipo === 'evidencia' && destino.tipo === 'evidencia') {
      errores.push({ codigo: 'CONEXION_INVALIDA', mensaje: `No se puede conectar dos nodos de evidencia directamente.`, aristaId: arista.id });
    }
  }

  return errores;
}

// ─── validarConectivo ──────────────────────────────────────

export function validarConectivo(
  conectivo: Conectivo,
  idsNodos: Set<Id>,
  idsConectivos?: Set<Id>,
): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];

  if (!conectivo.id) errores.push({ codigo: 'CONECTIVO_SIN_ID', mensaje: 'Conectivo sin identificador.' });
  if (idsConectivos?.has(conectivo.id)) errores.push({ codigo: 'CONECTIVO_DUPLICADO', mensaje: `Conectivo "${conectivo.id}" duplicado.`, conectivoId: conectivo.id });

  if (!conectivo.inputVariableIds || conectivo.inputVariableIds.length === 0) {
    errores.push({ codigo: 'CONECTIVO_SIN_INPUTS', mensaje: `Conectivo "${conectivo.id}" sin variables de entrada.`, conectivoId: conectivo.id });
  }
  if (!conectivo.outputVariableIds || conectivo.outputVariableIds.length === 0) {
    errores.push({ codigo: 'CONECTIVO_SIN_OUTPUTS', mensaje: `Conectivo "${conectivo.id}" sin variables de salida.`, conectivoId: conectivo.id });
  }

  for (const vid of [...conectivo.inputVariableIds, ...conectivo.outputVariableIds]) {
    if (!idsNodos.has(vid)) {
      errores.push({ codigo: 'CONECTIVO_VAR_INEXISTENTE', mensaje: `Conectivo "${conectivo.id}" referencia variable "${vid}" inexistente.`, conectivoId: conectivo.id });
    }
  }

  return errores;
}

// ─── validarMatriz ─────────────────────────────────────────

export function validarMatriz(
  matriz: Matriz,
  idsConectivos: Set<Id>,
  idsMatrices?: Set<Id>,
): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];

  if (!matriz.id) errores.push({ codigo: 'MATRIZ_SIN_ID', mensaje: 'Matriz sin identificador.' });
  if (idsMatrices?.has(matriz.id)) errores.push({ codigo: 'MATRIZ_DUPLICADA', mensaje: `Matriz "${matriz.id}" duplicada.`, matrizId: matriz.id });
  if (!idsConectivos.has(matriz.conectivoId)) errores.push({ codigo: 'MATRIZ_CONECTIVO_INEXISTENTE', mensaje: `Matriz "${matriz.id}" referencia conectivo "${matriz.conectivoId}" inexistente.`, matrizId: matriz.id });

  if (!matriz.valores || matriz.valores.length === 0) {
    errores.push({ codigo: 'MATRIZ_VACIA', mensaje: `Matriz "${matriz.id}" sin valores.`, matrizId: matriz.id });
  } else {
    const cols = matriz.valores[0]!.length;
    for (let i = 0; i < matriz.valores.length; i++) {
      if (matriz.valores[i]!.length !== cols) {
        errores.push({ codigo: 'MATRIZ_FILA_HETEROGENEA', mensaje: `Matriz "${matriz.id}" fila ${i} tiene ${matriz.valores[i]!.length} columnas, se esperaban ${cols}.`, matrizId: matriz.id });
      }
      for (let j = 0; j < matriz.valores[i]!.length; j++) {
        if (typeof matriz.valores[i]![j] !== 'number' || !Number.isFinite(matriz.valores[i]![j])) {
          errores.push({ codigo: 'MATRIZ_VALOR_INVALIDO', mensaje: `Matriz "${matriz.id}" [${i}][${j}] valor inválido.`, matrizId: matriz.id });
        }
      }
    }
  }

  return errores;
}

// ─── validarRed ─────────────────────────────────────────────

export function validarRed(red: Red): ErrorValidacion[] {
  const errores: ErrorValidacion[] = [];

  if (!red.id) errores.push({ codigo: 'RED_SIN_ID', mensaje: 'La red no tiene identificador.' });
  if (!red.nombre?.trim()) errores.push({ codigo: 'RED_SIN_NOMBRE', mensaje: 'La red no tiene nombre.' });

  if (!red.nodos || red.nodos.length === 0) {
    errores.push({ codigo: 'GRAFO_VACIO', mensaje: 'La red no contiene nodos.' });
    return errores;
  }

  const idsNodos = new Set<Id>();
  const idsAristas = new Set<Id>();
  const idsConectivos = new Set<Id>();
  const idsMatrices = new Set<Id>();

  for (const n of red.nodos) { errores.push(...validarNodo(n, idsNodos)); idsNodos.add(n.id); }
  for (const a of red.aristas) { errores.push(...validarArista(a, idsNodos, idsAristas, red.nodos)); idsAristas.add(a.id); }
  for (const c of red.conectivos) { errores.push(...validarConectivo(c, idsNodos, idsConectivos)); idsConectivos.add(c.id); }
  for (const m of red.matrices) { errores.push(...validarMatriz(m, idsConectivos, idsMatrices)); idsMatrices.add(m.id); }

  if (red.aristas.length > 0 && detectarCiclos(red.nodos, red.aristas)) {
    errores.push({ codigo: 'CICLO_DETECTADO', mensaje: 'La red contiene uno o más ciclos dirigidos.' });
  }

  return errores;
}